diffsmhm/diff_stats/cuda/sigma.py

```diff
@@ -0,0 +1,408 @@
+import numpy as np
+from numba import cuda
+
+import cmath
+
+
+def _copy_periodic_points_3D(x, y, z, boxsize, buffer_length):
+    # copy particles within buffer_length of an edge in XY
+    n_points = len(x)
+
+    # setup position copies
+    x_periodic = np.copy(x)
+    y_periodic = np.copy(y)
+    z_periodic = np.copy(z)
+
+    for p in range(n_points):
+        # if not near edge continue
+        if (
+            not (x[p] < buffer_length or x[p] > boxsize-buffer_length) and
+            not (y[p] < buffer_length or y[p] > boxsize-buffer_length) and
+            not (z[p] < buffer_length or z[p] > boxsize-buffer_length)
+        ):
+            continue
+
```

Comment:
This block could be written as a triple loop with three items for each dimension (left, middle, right) and skipping certain cases, but no need here. I like the explicit code.